cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,183 @@
+		}
+		allowedReplicasets = append(allowedReplicasets, replicaset)
+	}
+
+	return allowedReplicasets, nil
+}
+
+func Upgrade(replicasets replicaset.Replicasets, pendingReplicasetAliases []string,
+	lsnTimeout int) error {
+	allowedReplicasets, err := GetAllowedReplicasets(replicasets, pendingReplicasetAliases)
+	if err != nil {
+		return err
+	}
+
+	printReplicasetStatus := func(alias, status string) {
+		fmt.Printf("• %s: %s\n", alias, status)
+	}
+
+	for _, rs := range allowedReplicasets {
+		var masterRun *running.InstanceCtx = nil
+		var replicRun []running.InstanceCtx
+
+		for _, inst := range rs.Instances {
+			run := inst.InstanceCtx
+			fullInstanceName := running.GetAppInstanceName(run)
+			conn, err := connector.Connect(connector.ConnectOpts{
+				Network: "unix",
+				Address: run.ConsoleSocket,
+			})
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+
+			res, err := conn.Eval(
+				"return (type(box.cfg) == 'function') or box.info.ro",
+				[]any{}, connector.RequestOpts{})
+			if err != nil || len(res) == 0 {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+
+			if !res[0].(bool) {
+				if masterRun != nil {
+					printReplicasetStatus(rs.Alias, "error")
+					return fmt.Errorf("[%s]: %s and %s are both masters",
+						rs.Alias, running.GetAppInstanceName(*masterRun),
+						fullInstanceName)
+				}
+				masterRun = &run
+			} else {
+				replicRun = append(replicRun, run)
+			}
+		}
+		if masterRun == nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s]: has not master instance", rs.Alias)
+		}
+		var conn connector.Connector
+		conn, err = connector.Connect(connector.ConnectOpts{
+			Network: "unix",
+			Address: masterRun.ConsoleSocket,
+		})
+
+		if err != nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s][%s]: %s", rs.Alias,
+				running.GetAppInstanceName(*masterRun), err)
+		}
+		masterUpgradeInfo, err := internalUpgrade(conn)
+		if err != nil {
+			printReplicasetStatus(rs.Alias, "error")
+			return fmt.Errorf("[%s][%s]: %s", rs.Alias,
+				running.GetAppInstanceName(*masterRun), err)
+		}
+		for _, run := range replicRun {
+			fullInstanceName := running.GetAppInstanceName(run)
+			conn, err = connector.Connect(connector.ConnectOpts{
+				Network: "unix",
+				Address: run.ConsoleSocket,
+			})
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s][%s]: %s", rs.Alias, fullInstanceName, err)
+			}
+			err = WaitLSN(conn, masterUpgradeInfo, lsnTimeout)
+			if err != nil {
+				printReplicasetStatus(rs.Alias, "error")
+				return fmt.Errorf("[%s]: LSN wait timeout: error waiting LSN %d "+
+					"in vclock component %d on %s: time quota %d seconds "+
+					"exceeded", rs.Alias, masterUpgradeInfo.LSN,
+					masterUpgradeInfo.IID, fullInstanceName, lsnTimeout)
+			}
```

Comment:
The error returned from the LSN waiting function may be useful for a user. There may be different situations: we can't connect to the instance or the LSN just stucks.

I propose the following retrying behavior: each iteration saves its error and goes to the next one. When the timeout is reached, the error from the last iteration is returned.

This error is reported to the user together with the timeout error report. We can just add it at the end of the message.

Reply:
> We can just add it at the end of the message.

Done.